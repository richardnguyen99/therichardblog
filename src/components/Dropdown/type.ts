export type DropdownDirection = "top" | "bottom" | "left" | "right";

export interface DropdownProps {
  [key: string]: any; // For spreading variables
}

export interface DropdownContextValue {
  /**
   * @description Detects whether or not the dropdown is dropping or not
   */
  opening: boolean;

  /**
   * @description Inidicates which side of the toggle that the dropdown will be displayed
   *
   * @default "bottom";
   */
  direction?: DropdownDirection;

  /**
   * @description Controls the visibility of the dropdown component
   */
  toggle?: (nextState: boolean) => void;
}

export interface DropdownToggleProps {
  [key: string]: any; // For spreading variables
}

export interface DropdownMenuProps {
  [key: string]: any; // For spreading variables
}

export interface DropdownItemProps {
  /**
   * @description Fires a callback when the item is selected
   */
  onSelect?: () => void;

  [key: string]: any; // For spreading variables
}
