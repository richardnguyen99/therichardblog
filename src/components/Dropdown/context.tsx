/**
 * Context API for Dropdown components
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */
import * as React from "react";

import { DropdownContextValue } from "./type";

const DropdownContext = React.createContext<DropdownContextValue>({ opening: false });
DropdownContext.displayName = "DropdownContext";

export default DropdownContext;
