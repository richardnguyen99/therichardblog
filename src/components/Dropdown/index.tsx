/**
 * Core Dropdown component
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */
import * as React from "react";

import DropdownContext from "./context";
import { DropdownItemProps, DropdownMenuProps, DropdownProps, DropdownToggleProps } from "./type";
import {
  StyledDropdown,
  StyledDropdownItem,
  StyledDropdownMenu,
  StyledDropdownToggle,
} from "./style";
import useMergeRefs from "@hooks/useMergedRefs";

/**
 * @description Toggles to show or hide the dropdown menu component
 */
const DropdownToggle: React.FC<DropdownToggleProps> = React.forwardRef<
  HTMLButtonElement,
  DropdownToggleProps
>(({ children, ...rest }, ref) => {
  const dropdownCtx = React.useContext(DropdownContext);

  const handleClick = () => {
    dropdownCtx.toggle(!dropdownCtx.opening);
  };

  return (
    <StyledDropdownToggle ref={ref} onClick={handleClick} {...rest}>
      {children}
    </StyledDropdownToggle>
  );
});
DropdownToggle.displayName = "DropdownToggle";

/**
 * @description Display the collapsable content list
 */
const DropdownMenu: React.FC<DropdownMenuProps> = React.forwardRef<
  HTMLUListElement,
  DropdownMenuProps
>(({ children, ...rest }, ref) => {
  const dropdownCtx = React.useContext(DropdownContext);
  const innerRef = React.useRef<HTMLUListElement>(null);
  const mergedRef = useMergeRefs([innerRef, ref]);

  const handleClickOutside = React.useCallback(
    (evt) => {
      if (
        innerRef &&
        innerRef.current &&
        !innerRef.current.contains(evt.target) &&
        dropdownCtx.opening
      ) {
        dropdownCtx.toggle(false);
      }
    },
    [dropdownCtx, innerRef]
  );

  // Handle display animation
  React.useEffect(() => {
    if (innerRef && innerRef.current) {
      innerRef.current.style.display = dropdownCtx.opening && "block";
      innerRef.current.style.opacity = dropdownCtx.opening ? "1" : "0";
      innerRef.current.style.scale = dropdownCtx.opening ? "1" : "0.8";

      let interval;
      if (!dropdownCtx.opening) {
        interval = setInterval(() => {
          innerRef.current.style.display = "none";
        }, 1000);
      }

      return clearInterval(interval);
    }
  }, [dropdownCtx.opening, innerRef]);

  // Handle onClick event outside the component
  React.useEffect(() => {
    document.addEventListener("click", handleClickOutside);

    return () => {
      document.removeEventListener("click", handleClickOutside);
    };
  }, [handleClickOutside]);

  return (
    <StyledDropdownMenu ref={mergedRef} {...rest} style={{ opacity: 0 }}>
      {children}
    </StyledDropdownMenu>
  );
});
DropdownMenu.displayName = "DropdownMenu";

/**
 * @description Holds the item in the content
 */
const DropdownItem: React.FC<DropdownItemProps> = React.forwardRef<
  HTMLLIElement,
  DropdownItemProps
>(({ children, onSelect, ...rest }, ref) => {
  const dropdownCtx = React.useContext(DropdownContext);

  const handleClick = () => {
    if (onSelect) onSelect();

    dropdownCtx.toggle(false);
  };

  return (
    <StyledDropdownItem ref={ref} onClick={handleClick} {...rest}>
      {children}
    </StyledDropdownItem>
  );
});
DropdownItem.displayName = "DropdownItem";

/**
 * @description Renders a dropdown component that handles collapsable content
 */
const Dropdown: React.FC<DropdownProps> = React.forwardRef<HTMLDivElement, DropdownProps>(
  ({ children, ...rest }, ref) => {
    const [opening, setOpening] = React.useState(false);

    const toggle = React.useCallback(
      (nextShow: boolean) => {
        setOpening(nextShow);
      },
      [setOpening]
    );

    const contextValue = React.useMemo(
      () => ({
        opening,
        toggle,
      }),
      [opening, toggle]
    );

    return (
      <DropdownContext.Provider value={contextValue}>
        <StyledDropdown ref={ref} {...rest}>
          {children}
        </StyledDropdown>
      </DropdownContext.Provider>
    );
  }
);
Dropdown.displayName = "Dropdown";

export default Object.assign(Dropdown, {
  Toggle: DropdownToggle,
  Menu: DropdownMenu,
  Item: DropdownItem,
});
