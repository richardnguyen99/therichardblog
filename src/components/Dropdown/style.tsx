/**
 * Style definitions for dropdown components
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 */
import styled from "styled-components";

export const StyledDropdown = styled.div`
  position: relative;
`;

export const StyledDropdownMenu = styled.ul`
  position: absolute;
  top: 100%;
  left: 0;
  z-index: 100;
  width: 160px;
  margin-top: 0.125rem;
  list-style: none;
  background-color: #18181b; // TODO: Implement the background base for light theme
  color: #efeff1; // TODO: Implement the color bas forr light theme
  border-radius: 0.6rem;
  box-shadow: 0 12px 32px rgba(0, 0, 0, 0.5), 0 0px 8px rgba(0, 0, 0, 0.4);
  padding: 0.5rem;
  transition: all 0.1s ease;

  &::before,
  &::after {
    position: absolute;
    display: inline-block;
    content: "";
  }

  &::before {
    border: 8px solid transparent;
    border-bottom-color: #18181b;
  }

  &::after {
    border: 7px solid transparent;
    border-bottom-color: #18181b;
  }

  &.south {
    right: 50%;
    left: auto;
    transform: translateX(50%);

    &::before {
      top: -16px;
      right: 50%;
      transform: translateX(50%);
    }

    &::after {
      top: -14px;
      right: 50%;
      transform: translateX(50%);
    }
  }

  &.southwest {
    right: 0;
    left: auto;

    &::before {
      top: -16px;
      right: 9px;
      left: auto;
    }

    &::after {
      top: -14px;
      right: 10px;
      left: auto;
    }
  }
`;

export const StyledDropdownToggle = styled.button`
  position: relative;
  user-select: none;
  cursor: pointer;
  z-index: 1;
  overflow: hidden;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
  vertical-align: middle;
  background: transparent;
  fill: #efefe1;
  color: #efefe1;
  width: 2rem;
  height: 2rem;
  margin: 4px;
  padding: 16px;
  border-radius: 0.3rem;
  border: none;
  transition: all 0.25s ease-in 10ms;
  &:hover {
    background: rgba(255, 255, 255, 0.2);
  }
  &:active {
    background: rgba(255, 255, 255, 0.15);
  }
`;

export const StyledDropdownItem = styled.li`
  display: block;
  padding: 4px 8px 4px 16px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  border-radius: 0.25rem;
  &:hover {
    background: rgba(255, 255, 255, 0.2);
    color: #efefe1;
  }
`;
