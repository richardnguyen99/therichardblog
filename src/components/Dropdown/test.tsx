/**
 * Core test file for dropdown component
 *
 * @author Richard Nguyen <richard.ng0616@gmail.com>
 *
 * @jest-environment jsdom
 */
import * as React from "react";
import { render } from "@testing-library/react";

import Dropdown from "./";

describe("<Dropdown>", () => {
  it("should match snapshot", () => {
    const { container } = render(
      <Dropdown>
        <Dropdown.Toggle style={{ padding: 0 }}>
          <svg
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 16 16"
            width="16"
            height="16"
            fill="white"
            fillRule="evenodd">
            <path d="M8 9a1.5 1.5 0 100-3 1.5 1.5 0 000 3zM1.5 9a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm13 0a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"></path>
          </svg>
        </Dropdown.Toggle>
        <Dropdown.Menu className="southwest">
          <Dropdown.Item>Tags</Dropdown.Item>
          <Dropdown.Item>Tags</Dropdown.Item>
          <Dropdown.Item>Tags</Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    );

    expect(container).toMatchSnapshot();
  });
});
