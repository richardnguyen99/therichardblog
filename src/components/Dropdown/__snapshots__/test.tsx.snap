// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown> should match snapshot 1`] = `
<div>
  <div
    class="sc-bdfBQB lfCfsQ"
  >
    <button
      class="sc-dlfnuX jJZYL"
      style="padding: 0px;"
    >
      <svg
        fill="white"
        fill-rule="evenodd"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 9a1.5 1.5 0 100-3 1.5 1.5 0 000 3zM1.5 9a1.5 1.5 0 100-3 1.5 1.5 0 000 3zm13 0a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
        />
      </svg>
    </button>
    <ul
      class="sc-gsTEea gqUDwU southwest"
      style="opacity: 0;"
    >
      <li
        class="sc-hKgJUU yprVz"
      >
        Tags
      </li>
      <li
        class="sc-hKgJUU yprVz"
      >
        Tags
      </li>
      <li
        class="sc-hKgJUU yprVz"
      >
        Tags
      </li>
    </ul>
  </div>
</div>
`;
